### PR TITLE
Add fuel prices

### PR DESCRIPTION
**Pull request by IntegratedQuantum**
At each planet the player pays for refueling. The price for this can be changed for each the government.
I've given all governments a value of 1 except for high developed government which got 0.5 or 0 and pirate/piratelike and economy governments which want 2

(planned) changes:
- [x] The player has to pay a government specific fuel price when being refueled.
- [x] Make planets independent of the government prices
- [x] add auto-refuel preferences
- [x] include an interface where the player can refuel the ships
- [x] make can refuel and hasSpaceport independent
- [x] make escorts pay fuel price when landed
- [x] make escorts not refuel if price is to expensive
- [x] include support for changes created by events
- [x] add refuel help
- [x] hide the bar if the fleet is fully refueled
- [x] resolve conflicts

### CODE REVIEW

**Comment by finite-galaxy** [comment]
Nice! The idea to have government-specific prices is great as well! Although I haven't tried it out yet, here are some initial remarks:
* what is "refillCrew"?
* "tons of fuel" → "units of fuel"
* make fuel price the cost of 1 unit of fuel, because you'll not always need multiples of 100, and because ships can have odd capacities (have a look at Mereti and Sestor ships)
  * what happens if the fuel price is not an int? E.g. 1.23
  * what happens if no fuel price is specified? Will those ships give fuel for free? Or will they simply refuse to refuel you?
* make refuelling on a planet optional; perhaps a pop-up message upon landing *and* departure? Some captains might prefer not to pay for fuel and rely on ramscoop instead
  * what happens if you buy a new ship? Is fuel included?
  * what happens if you buy or install a fuel pod?
* how much work would it be to add support for planet-specific prices? E.g. I can imagine a Republic ships charges you 2 ¢/fuel in space, but you can buy it at 1 ¢/fuel on a populated world, at 0.5 ¢/fuel on a fuel refinery station, and at 1.5 ¢/fuel on an isolated moon. (This might belong in a separate patch.)

**Comment by IntegratedQuantum** [comment]
>     * what is "refillCrew"?

On landing the crew is restocked, but I didn't want to add refuel while landing(choosing departure is in my opinion better because I wouldn't want to pay the price for refueling my ship if I was going to sell it afterwards) and I don't know of a better way to do this.

>     * make fuel price the cost of 1 unit of fuel, because you'll not always need multiples of 100, and because ships can have odd capacities (have a look at Mereti and Sestor ships)

I didn't want to have non-integer prices, but if you want i can change it. Also it is no problem to have odd capacities because the total price (fuel*priceof100) is divided by 100 and not the number of fuel units.

>       * what happens if no fuel price is specified? Will those ships give fuel for free? Or will they simply refuse to refuel you?

They will give fuel for free. It would seem unnatural for me to explicitly assign your escort a fuel price( of 0) which would need to be done for another standard price.

>     * make refuelling on a planet optional; perhaps a pop-up message upon landing _and_ departure? Some captains might prefer not to pay for fuel and rely on ramscoop instead

Some might prefer not paying for fuel, but others might prefer not having to click on the screen on every departure.
It might be possible to add an outfit that let's your ship orbit the planet and collect fuel while you are on the ground.

>       * what happens if you buy a new ship? Is fuel included?
>       * what happens if you buy or install a fuel pod?

They are sold with full fuel. 

>     * how much work would it be to add support for planet-specific prices? E.g. I can imagine a Republic ships charges you 2 ¢/fuel in space, but you can buy it at 1 ¢/fuel on a populated world, at 0.5 ¢/fuel on a fuel refinery station, and at 1.5 ¢/fuel on an isolated moon.

It would be a lot of work inserting them for every planet, but if you would make the change only on a few worlds this might be possible.

>     (This might belong in a separate patch.)

Yes.

**Comment by IntegratedQuantum** [comment]
If calculated on departure there is again a problem with the map.

**Comment by finite-galaxy** [comment]
Basically what I want:
* when hailing ships in space, you obtain jump fuel - current fuel at *government* fuel price
  * message: “You received m fuel and paid n credits as compensation.”
* when on a planet, you get fuel capacity - current fuel at *planet* fuel price
  * pop-up: [x fuel at y cost xy credits to completely refuel your fleet][Cancel][OK]

Defaulting to 0 if no fuel price is specified is fine; you don't have to add it to every planet and government, I can do that later, just add it to a few to show it works (e.g. Earth 1.0, Mars 1.2, Luna 1.5)

Also, I prefer fuel prices for single units, because 1.5 is clearer than 150/100. Besides, most plug-in writers will only look at the data files, not at the source code.

Instead of a pop-up window on arrival or departure, it might be better to have a Refuel button (or press R), directly above Depart. What do you think?

To check:
* What happens if you install a fuel pod from cargo?
* What happens if you sell an installed fuel pod and subsequently re-add them?
* Do you pay for the fuel if you purchase a fuel pod?
* Do you get paid for the fuel if you sell a ship?

**Comment by IntegratedQuantum** [comment]
> Basically what I want:
> 
>     * when hailing ships in space, you obtain jump fuel - current fuel at _government_ fuel price
>       
>       * message: “You received m fuel and paid n credits as compensation.”

That's the way it's currently working(except the text is a bit different).

>     * when on a planet, you get fuel capacity - current fuel at _planet_ fuel price
>       
>       * pop-up: [x fuel at y cost xy credits to completely refuel your fleet][Cancel][OK]
> 
> Instead of a pop-up window on arrival or departure, it might be better to have a Refuel button (or press R), directly above Depart. What do you think?

I don't really like both options:
1. You would would have to press 'R' or 'Enter' on every planet to refuel.
2. Using a pop-up display every player would have to press a button(even if they don't want to refuel) while with 'R' most players(especially new players) would forget to refuel.
3. You would have no control about what is happening with your escort(example: your flagship has a ramscoop while your escort doesn't, so you don't want to refuel your flagship, but your escort really needs fuel.).
4. What happens if your escort is out of fuel in a distant system and has no ramscoop?
5. I don't think it's worth the effort to have every player click thousands of times just so that a few players can pay a few credits less for their fuel(and it really is just a tiny sum compared to the crew costs).

One way of solving all these problems could be having your refuel button which leads to a display where you can decide to refuel only if the local price is under a certain price and only refuel those of your ships which have a certain ramscoop/effectiveFuel. Those to values would be always refuel as standard value. For the rare occasions where the player's rule doesn't apply(for example he/she/it wants to refuel all ships on a planet with extremely low prices) there could be the response to automatically refuel when pressing r. Also there might be an option to sell all fuel and to gradually refuel all ships to be prepared for n jumps afterwards.

> Defaulting to 0 if no fuel price is specified is fine; you don't have to add it to every planet and government, I can do that later, just add it to a few to show it works (e.g. Earth 1.0, Mars 1.2, Luna 1.5)

It might be a bit easier if the system was using the government-specific value if not stated otherwise(so you wouldn't have to assign every planet a value).

> Also, I prefer fuel prices for single units, because 1.5 is clearer than 150/100. Besides, most plug-in writers will only look at the data files, not at the source code.

Ok, I'll change that.

**Comment by finite-galaxy** [comment]
> One way of solving all these problems could be having your refuel button which leads to a display where you can decide to always automatically refuel if the local price is under a certain price and only refuel those of your ships which have a certain ramscoop/effectiveFuel(or something with similar meaning). Those to values would be always refuel as standard value. For the rare occasions where the player's rule doesn't apply(for example he/she/it wants to refuel all ships on a planet with extremely low prices) there could be the response to automatically refuel when pressing r. Also there might be an option to sell all fuel and to gradually refuel all ships to be prepared for n jumps afterwards.
Of course the ships would always refuel when the fuel costs are 0.

Sounds good, although it's probably more work to implement that.

> It might be a bit easier if the system was using the government-specific value if not stated otherwise(so you wouldn't have to assign every planet a value).

No, planets shouldn't use universal government fuel prices. If a planet has a spaceport, ships can refuel there, and if no fuel price is specified for that planet, it is 0. (Don't worry about me.)

> This branch has conflicts that must be resolved

**Comment by IntegratedQuantum** [comment]
> Sounds good, although it's probably more work to implement that.

A better solution is not always easy.

> No, planets shouldn't use universal government fuel prices. If a planet has a spaceport, ships can refuel there, and if no fuel price is specified for that planet, it is 0. (Don't worry about me.)

I'm not worrying about you, but about plug-in writers who will think that all planets usually use the government's price(which is an intuitive way of perceiving a government-specific fuel price). Also I don't think they would want to spend quite some time to give every planet a non-zero fuel price(after some time of finding out why there is no fuel price for their new planets).
Also map.txt would be slightly smaller than in your case.

> > This branch has conflicts that must be resolved

I'll care about that when I'm done with everything.

**Comment by finite-galaxy** [comment]
> I'm not worrying about you, but about plug-in writers who will think that all planets usually use the government's price(which is an intuitive way of perceiving a government-specific fuel price).

Not all planets have bribe, required reputation, or security values specified either, which is not a problem, nor do those default to government-specific values; *mutatis mutandis* for fuel price, which should be planet specific as well.

**Comment by IntegratedQuantum** [comment]
> Not all planets have bribe, required reputation, or security values specified either, which is not a problem, nor do those default to government-specific values;

The main difference between those and fuel prices is that they have no default value at all because not assigning them results in no behaviour at all(which is ok if you don't want to give a planet the ability to be bribed or have any security) while giving fuel for free doesn't seem right if someone just forgot to assign a value.

**Comment by finite-galaxy** [comment]
> The main difference between those and fuel prices is that they have no default value at all because not assigning them results in no behaviour at all(which is ok if you don't want to give a planet the ability to be bribed or have any security) while giving fuel for free doesn't seem right if someone just forgot to assign a value.

Now you mention it, perhaps it's better to separate fuel from the spaceport, i.e. ships can only refuel on planets which have a fuel price specified. This gives more flexibility (planets can then have: i. fuel and spaceport, ii. spaceport but no fuel, iii. fuel but no spaceport, iv. neither; instead of either both or neither).
 How much work would that be to add?

> Should I make the refuel preferences an option that is stored for each specific game or for every game(like it's done for all other options)?

You mean the preference option where you can decide whether you ships:
* always refuel, regardless of fuel price
* automatically refuel is fuel price is x or less
* pop-up a message on every planet, showing the local fuel price, allowing you to manually decide whether or not to refuel
* never refuel

For now I think it's best to save it similar to e.g. “Escorts expend ammo” preference. In the future the settings might be split into a set specific to a single pilot and a set applying to all pilots, but that is beyond the scope of this pull request.

**Comment by IntegratedQuantum** [comment]
>     Now you mention it, perhaps it's better to separate fuel from the spaceport, i.e. ships can only refuel on planets which have a fuel price specified. This gives more flexibility (planets can then have: i. fuel and spaceport, ii. spaceport but no fuel, iii. fuel but no spaceport, iv. neither; instead of either both or neither).
How much work would that be to add?

Having planets that don't refuel is a good idea, especially to have alien governments who only rely on ramscoops or use entirely different fuel.

>     * pop-up a message on every planet, showing the local fuel price, allowing you to manually decide whether or not to refuel
> 
>     * never refuel

I don't want to add these 2 options because you can refuel in the refuelpanel on each planet and it would be stupid to never refuel even if it's for free.

**Comment by finite-galaxy** [comment]
> Having planets that don't refuel is a good idea, especially to have alien governments who only rely on ramscoops or use entirely different fuel.

Yes, the Wanderers perhaps.

> I don't want to add these 2 options because you can refuel in the refuelpanel on each planet and it would be stupid to never refuel even if it's for free.

So there will be a refuel panel on planets, similar to the hire crew panel, where you can see the local fuel price and manually decide whether or not to refuel? Great! Then the auto-refuel preference is simply a nice to have bonus, ignore those last two options; go ahead.

**Comment by finite-galaxy** [comment]
Could you update your opening post to summarize what's been changed and lists what still has to be done?

Also, events should be able to add, remove, or change a planet's fuel price (could be done here or in a separate patch).

**Comment by IntegratedQuantum** [comment]
I have added the interface now:
screenshot from 2019-01-17 21-44-24

Other possible texts below the bars are "Only refuels if it is for free"; "Refuel independently of the price" and "Refuel independently of ramscoop value of the ship".

I'm not sure about the scaling of the second bar. What should maximum value for ramscoop/jump fuel be?

**Comment by finite-galaxy** [comment]
Is that a slide-bar?

I assumed the auto-refill option would go under Preferences/Settings.
Also, I don't understand the purpose of a ramscoop per jump fuel value.

The refuel panel should have probably have the following:
* The fuel price on this planet is x credits
* Refuel your ships up to w%; which will cost y credits in total
* Completely refuelling your entire fleet will cost z credits in total
Perhaps it's worth trying organizing it in a tabular form, similar to the hire crew panel?

**Comment by finite-galaxy** [comment]
Is that a slide-bar?

I assumed the auto-refill option would go under Preferences/Settings.
Also, I don't understand the purpose of a ramscoop per jump fuel value.

The refuel panel should have probably have the following:
* The fuel price on this planet is x credits
* Refuel your ships up to w%; which will cost y credits in total
* Completely refuelling your entire fleet will cost z credits in total

Perhaps it's worth trying organizing it in a tabular form, similar to the hire crew panel?

**Comment by IntegratedQuantum** [comment]
> Is that a slide-bar?

It's the same bar that you can use to regulate audio volume in the preferences.

> I assumed the auto-refill option would go under Preferences/Settings.

I thought it would be useful to have it in the refuelpanel because that's where i would search for such an option. But I could additionally add this to the Preferences/settings

> Also, I don't understand the purpose of a ramscoop per jump fuel value.

Ramscoop/jump fuel basically gives you information about what part of your jump fuel is refueled in one time unit.
I'm not completely sure what exactly the ramscoop value of a ramscoop is. If I knew that I could calculate the average time a ship needs in space to recharge the fuel for one jump which could be understood better by the player than what i have now.

**Comment by finite-galaxy** [comment]
>I thought it would be useful to have it in the refuelpanel because that's where i would search for such an option. But I could additionally add this to the Preferences/settings

Well, I assumed it was a setting that makes your fleet automatically refuel when landing on a planet, without you having to access the refuel panel or clicking anything; options belong under preferences/settings.

>Ramscoop/jump fuel basically gives you information about what part of your jump fuel is refueled in one time unit.
I'm not completely sure what exactly the ramscoop value of a ramscoop is. If I knew that I could calculate the average time a ship needs in space to recharge the fuel for one jump which could be understood better by the player than what i have now.

The fuel gain from ramscoop depends on the sqrt(ramscoop), the system's solar wind, and the ship's distance to the system's centre (see Ship::DoGeneration). In other words, it's not a reliable constant (fuel generation is), therefore it's probably better to scrap the ramscoop/jump fuel proposal.

**Comment by IntegratedQuantum** [comment]
> Well, I assumed it was a setting that makes your fleet automatically refuel when landing on a planet, without you having to access the refuel panel or clicking anything; options belong under preferences/settings.

My problem about putting this only to the preferences is that new players might not realise there actually is such an option.

> The fuel gain from ramscoop depends on the sqrt(ramscoop), the system's solar wind, and the ship's distance to the system's centre (see Ship::DoGeneration). In other words, it's not a reliable constant (fuel generation is), therefore it's probably better to scrap the ramscoop/jump fuel proposal.

That's why I would want to calculate some kind of average value which would at least give you the option to decide a few of your ships won't need to refuel because they recharge fuel on average fast enough. Otherwise I would need to make the option to manually refuel single ships which I wouldn't want to implement.

**Comment by finite-galaxy** [comment]
> My problem about putting this only to the preferences is that new players might not realise there actually is such an option.

Which is the case for practically all options.

> That's why I would want to calculate some kind of average value which would at least give you the option to decide a few of your ships won't need to refuel because they recharge fuel on average fast enough. 

Yeah, but ramscoop is too unreliable.

> Otherwise I would need to make the option to manually refuel single ships which I wouldn't want to implement.

No, manually refuelling individual ships is a hell for anyone with more than three ships. What could be done is adding a “refuel your fleet up to x times jumpfuel” choice, but then there might be that cargo issue again.

**Comment by IntegratedQuantum** [comment]
> but then there might be that cargo issue again.

Why would it be there again?

**Comment by finite-galaxy** [comment]
> Why would it be there again?

If you partially refuel your ships first and buy cargo afterwards?

**Comment by IntegratedQuantum** [comment]
> If you partially refuel your ships first and buy cargo afterwards?

Then the player should refuel after buying stuff. And I might add a note that this is relevant.

**Comment by finite-galaxy** [comment]
Perhaps put that option at the bottom of “other” section in the second column, which also lists the rehire crew option? It might also be better to expand the sentence to: “Automatically refuel fleet on planets if fuel price is at or below x”.

**Comment by finite-galaxy** [comment]
Another thing to check: are credits deduced if escorts run out of fuel and land on a planet while your flagship stays in space?

**Comment by IntegratedQuantum** [comment]
Not yet.

**Comment by IntegratedQuantum** [comment]
screenshot from 2019-01-28 22-43-32

Now I changed the interface so that a certain ration of the fuel tanks can be filled.

**Comment by IntegratedQuantum** [comment]
I guess I'm done now.

**Comment by finite-galaxy** [comment]
When I tried compiling it (`scons`) I encountered the following errors:
```
code/PlayerInfo.cpp: In member function ‘void PlayerInfo::Land(UI*)’:
code/PlayerInfo.cpp:1044:8: error: redeclaration of ‘bool hasSpaceport’
   bool hasSpaceport = planet->HasSpaceport() && planet->CanUseServices();
        ^~~~~~~~~~~~
code/PlayerInfo.cpp:1008:8: note: ‘bool hasSpaceport’ previously declared here
   bool hasSpaceport = planet->HasSpaceport() && planet->CanUseServices();
        ^~~~~~~~~~~~
code/PlayerInfo.cpp: In member function ‘void PlayerInfo::RefuelRatio(double)’:
code/PlayerInfo.cpp:1408:30: error: qualified-id in declaration before ‘(’ token
 double PlayerInfo::FuelNeeded(double ratio)
                              ^
code/PlayerInfo.cpp:2813:1: error: expected ‘}’ at end of input
 }
 ^
code/PlayerInfo.cpp:1398:3: note: to match this ‘{’
   {
   ^
code/PlayerInfo.cpp:2813:1: error: expected ‘}’ at end of input
 }
 ^
code/PlayerInfo.cpp:1388:1: note: to match this ‘{’
 {
 ^
scons: *** [build/release/PlayerInfo.o] Error 1
scons: building terminated because of errors.
```

**Comment by IntegratedQuantum** [comment]
It seems like I made some mistakes while resolving conflicts.

**Comment by finite-galaxy** [comment]
Thank you, it compiles without errors now. I briefly tried it out, jumping around a couple of systems and subsequently landing. This is what it looked like:
screenshot from 2019-02-09 21-05-24

What you can see:
* my ships automatically refuelled upon landing (default setting?)
* there was no message informing me how much it has cost (should be below “Landing on <planet>”
* refuel panel still shows my ships can refuel, yet “Refuel” button does nothing
* font size is visibly smaller than other panels

Also, I'd prefer governments have a "fuel price" (or "fuel cost") value, but planets simply "fuel" (cf. security, tribute).

File: code/PlayerInfo.cpp [hunk 1]
```diff
@@ -1342,6 +1345,78 @@ void PlayerInfo::UnloadCargo()
...
+        ship->Recharge(false);
+    }
+  
+  // Pays the fuel and messages the price that was paid.
+  if(rechargedFuel && price > 0)
+  {
+    int fuelPrice = rechargedFuel*price;
+    ostringstream out;
+    out << "You paid " << fuelPrice << " credits to buy " << rechargedFuel << " units of fuel.";
+    Messages::Add(out.str());
```
Comment (IntegratedQuantum) [inline code/PlayerInfo.cpp]: As you can see here there should be a message. Have you changed the messaging system in the last weeks?

**Comment by IntegratedQuantum** [comment]
Also there is a help that tells you what the default setting is.

**Comment by finite-galaxy** [comment]
It seems NPC ships run out of fuel:
screenshot from 2019-02-10 13-51-20

**Comment by IntegratedQuantum** [comment]
That should fix it.

**Comment by finite-galaxy** [comment]
Thanks. Two more minor points:
* refuel panel body text is white; it ought to be light grey; cf. other panels
* perhaps rephrase the refuelling-in-space message to “You paid NPC x credits for refuelling your flagship” or something; fractional amounts of fuel don't make much sense:
screenshot from 2019-02-10 19-46-54

**Comment by finite-galaxy** [comment]
Because it's a large patch, I did some additional play-testing, trying to think of things that might go wrong. Something that should be changed is the behaviour of NPC escorts (ships that you don't own); they should ignore your fuel preference and always refuel on planets, independent of fuel price, whenever they run out of fuel.
To reproduce test: accept one or more escort jobs, set fuel price preference to 0, travel into a different direction, leaving your escort behind, and observe it being stranded in space.

**Comment by IntegratedQuantum** [comment]
Ok that fixes it.
I also found and fixed another problem:
when you landed and an escort was landed on another planet, after depart it wasn't fully refueled and had to pay again the refuel price.

**Comment by finite-galaxy** [comment]
Thank you for the patch! It works, seems complete, and I can't think of things that could go wrong right now, therefore I've merged it.

**Pull request merged**